File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_lcd2_test.ts
/**
 * g_bios_lcd_text（CPU→IO コマンド 18h）
 * 根拠: HandShake.mdc「LCD文字列表示」/ boot_monitor.mdc / test_framework.mdc
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

/** 文字列バッファ（ワードアドレス、1ワード1バイト） */
const TEXT_BUF = 0x1a00;

const BASE_REGS = {
  R4: 0x4444,
} as const;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ ioMock handshake で 1 ケースを実行する。
 * @param fn 本体（session / mock 利用可）
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * ワードバッファへ 1 バイト／ワードで書く。
 * @param s セッション
 * @param wordAddr 先頭ワードアドレス
 * @param bytes 下位 8bit ずつ
 */
function writeByteWords(
  s: Mn1613AsmSession,
  wordAddr: number,
  bytes: readonly number[],
): void {
  for (let i = 0; i < bytes.length; i += 1) {
    s.writeWord(wordAddr + i, bytes[i]! & 0xff);
  }
}

/**
 * g_bios_lcd_text を呼び、CPU→IO 1 トランザクションと並行する。
 * @param mock IO モック
 * @param row 行（R0 Bit8–9）
 * @param col 列（R0 Bit0–7）
 * @param len 文字数（R1）
 * @param textAddr 文字列先頭（R2）
 */
async function callLcd2(
  mock: IoBoardHandshakeMock,
  row: number,
  col: number,
  len: number,
  textAddr: number,
): Promise<void> {
  await Promise.all([
    session.call("g_bios_lcd_text", {
      registers: {
        ...BASE_REGS,
        R0: ((row & 0x3) << 8) | (col & 0xff),
        R1: len,
        R2: textAddr,
        R3: 0x3333,
      },
    }),
    mock.handleOneRequest(),
  ]);
}

test("18h フレームを20バイトで送り、len以降は空白で埋める", async () => {
  await withCase(async (s, mock) => {
    writeByteWords(s, TEXT_BUF, [0x41, 0x42, 0x43, 0x44, 0x45]); // ABCDE

    await callLcd2(mock, 1, 2, 5, TEXT_BUF);

    const last = mock.state.log.at(-1);
    expect(last).toBeTruthy();
    expect(last!.cmd).toBe(0x18);
    expect(last!.frame.length).toBe(20);
    expect(Array.from(last!.frame.slice(0, 4))).toEqual([
      0x18, 0x01, 0x02, 0x05,
    ]);
    expect(Array.from(last!.frame.slice(4, 9))).toEqual([
      0x41, 0x42, 0x43, 0x44, 0x45,
    ]);
    expect(Array.from(last!.frame.slice(9))).toEqual(new Array(11).fill(0x20));

    const status = last!.response?.[0] ?? 0;
    s.expectRegisters({ R0: status & 0xff });
  });
});

test("len が 16 を超えると先頭16文字だけ送る", async () => {
  await withCase(async (s, mock) => {
    const chars = Array.from({ length: 18 }, (_, i) => 0x41 + i); // A..R
    writeByteWords(s, TEXT_BUF, chars);

    await callLcd2(mock, 0, 0, 18, TEXT_BUF);

    const last = mock.state.log.at(-1);
    expect(last).toBeTruthy();
    expect(last!.cmd).toBe(0x18);
    expect(Array.from(last!.frame.slice(4))).toEqual(chars.slice(0, 16));
  });
});

test("R3/R4 は呼び出しの前後で保たれる", async () => {
  await withCase(async (s, mock) => {
    writeByteWords(s, TEXT_BUF, [0x48, 0x49]); // HI
    await callLcd2(mock, 0, 0, 2, TEXT_BUF);
    s.expectRegisters({ R3: 0x3333, R4: BASE_REGS.R4 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_lcd2_test.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

const TEXT_BUF: u16 = 0x1a00;

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn collect_cpu_frame(
    mock: &IoBoardHandshakeMock,
    expected_len: usize,
) -> Result<Vec<u8>, FrameworkError> {
    let mut frame = Vec::with_capacity(expected_len);
    let start = Instant::now();
    while frame.len() < expected_len {
        if let Some(chunk) = mock.take_cpu_to_io_frame() {
            frame.extend(chunk);
            continue;
        }
        if start.elapsed() > Duration::from_millis(2000) {
            return Err(FrameworkError::invalid_argument(
                "timeout collecting cpu_to_io frame",
            ));
        }
        std::thread::yield_now();
    }
    frame.truncate(expected_len);
    Ok(frame)
}

fn wait_in_dack<P>(
    mock: &IoBoardHandshakeMock,
    timeout: Duration,
    pred: P,
) -> Result<(), FrameworkError>
where
    P: Fn(u8) -> bool,
{
    let start = Instant::now();
    loop {
        let dack = mock.wires.lock().expect("wires lock").hshk_in_dack;
        if pred(dack) {
            return Ok(());
        }
        if start.elapsed() > timeout {
            return Err(FrameworkError::invalid_argument(
                "timeout waiting hshk_in_dack",
            ));
        }
        std::thread::yield_now();
    }
}

fn feed_io_response(mock: &IoBoardHandshakeMock, data: &[u8]) -> Result<(), FrameworkError> {
    let timeout = Duration::from_millis(2000);
    {
        let mut w = mock.wires.lock().expect("wires lock");
        w.hshk_in_req = 1;
        w.hshk_in_dena = 0;
    }
    let mut i = 0usize;
    while i < data.len() {
        let b0 = data[i];
        let b1 = if i + 1 < data.len() { data[i + 1] } else { 0 };
        {
            let mut w = mock.wires.lock().expect("wires lock");
            w.hshk_in_data = b0;
            w.hshk_in_dena = 1;
        }
        wait_in_dack(mock, timeout, |d| d != 0)?;
        {
            let mut w = mock.wires.lock().expect("wires lock");
            w.hshk_in_data = b1;
            w.hshk_in_dena = 0;
        }
        wait_in_dack(mock, timeout, |d| d == 0)?;
        i += 2;
    }
    mock.wires.lock().expect("wires lock").hshk_in_req = 0;
    Ok(())
}

fn serve_cpu_to_io(
    mock: &IoBoardHandshakeMock,
    expected_len: usize,
) -> Result<Vec<u8>, FrameworkError> {
    let req = collect_cpu_frame(mock, expected_len)?;
    let reply = mock.dispatch_cpu_to_io(&req);
    feed_io_response(mock, &reply)?;
    Ok(req)
}

fn write_byte_words(s: &mut Mn1613AsmSession, word_addr: u16, bytes: &[u8]) {
    for (i, b) in bytes.iter().enumerate() {
        s.write_word(word_addr.wrapping_add(i as u16), (*b as u16) & 0x00ff);
    }
}

fn call_lcd2(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    row: u16,
    col: u16,
    len: u16,
    text_addr: u16,
) -> Result<Vec<u8>, FrameworkError> {
    let io = Arc::clone(mock);
    let worker = std::thread::spawn(move || serve_cpu_to_io(&io, 20));

    let _ = s.call(
        "g_bios_lcd_text",
        CallOptions {
            registers: Some(CallRegisters {
                r0: Some(((row & 0x03) << 8) | (col & 0xff)),
                r1: Some(len),
                r2: Some(text_addr),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            }),
            ..Default::default()
        },
    )?;

    let req = worker
        .join()
        .map_err(|_| FrameworkError::invalid_argument("io worker panicked"))??;
    Ok(req)
}

#[test]
fn lcd_text_sends_20byte_frame_and_pads_spaces() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        write_byte_words(s, TEXT_BUF, &[0x41, 0x42, 0x43, 0x44, 0x45]);
        let req = call_lcd2(s, &mock, 1, 2, 5, TEXT_BUF)?;

        assert_eq!(req.len(), 20);
        assert_eq!(&req[0..4], &[0x18, 0x01, 0x02, 0x05]);
        assert_eq!(&req[4..9], &[0x41, 0x42, 0x43, 0x44, 0x45]);
        assert_eq!(&req[9..20], &[0x20; 11]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn lcd_text_len_gt_16_sends_first_16_chars_only() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let chars: Vec<u8> = (0..18).map(|i| 0x41 + i as u8).collect();
        write_byte_words(s, TEXT_BUF, &chars);
        let req = call_lcd2(s, &mock, 0, 0, 18, TEXT_BUF)?;

        assert_eq!(&req[4..20], &chars[0..16]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn lcd_text_preserves_r3_r4() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        write_byte_words(s, TEXT_BUF, &[0x48, 0x49]);
        let _ = call_lcd2(s, &mock, 0, 0, 2, TEXT_BUF)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
